cpp/src/svm/results.h

```diff
@@ -0,0 +1,311 @@
+    get_rows<<<MLCommon::ceildiv(n_support * n_cols, TPB), TPB, 0, stream>>>(
+      x, n_rows, n_cols, x_support, n_support, idx);
+    CUDA_CHECK(cudaPeekAtLastError());
+    return x_support;
+  }
+
+  /* Calculate coefficients = alpha * y
+   * @param n_support number of support vertors
+   * @param alpha
+   * @return buffer with dual coefficients, size [n_support]
+   */
+  math_t *GetDualCoefs(int n_support, const math_t *alpha) {
+    auto allocator = handle.getDeviceAllocator();
+    math_t *dual_coefs =
+      (math_t *)allocator->allocate(n_support * sizeof(math_t), stream);
+    MLCommon::device_buffer<math_t> math_tmp(allocator, stream, n_rows);
+    // Calculate dual coefficients = alpha * y
+    MLCommon::LinAlg::binaryOp(
+      math_tmp.data(), alpha, y, n_rows,
+      [] __device__(math_t a, math_t y) { return a * y; }, stream);
+    // Return only the non-zero coefficients
+    auto select_op = [] __device__(math_t a) { return 0 < a; };
+    SelectByAlpha(alpha, n_rows, math_tmp.data(), select_op, dual_coefs);
+    return dual_coefs;
+  }
+
+  /**
+   * Flag support vectors and also collect their indices.
+   * Support vectors are the vectors where alpha > 0.
+   *
+   * @param [in] alpha dual coefficients, size [n_rows]
+   * @param [out] n_support number of support vectors
+   * @param [out] idx indices of the suport vectors, size [n_support]
+   */
+  void GetSupportVectorIndices(const math_t *alpha, int *n_support, int **idx) {
+    *n_support = SelectByAlpha(
+      alpha, n_rows, f_idx.data(),
+      [] __device__(math_t a) -> bool { return 0 < a; }, idx_selected.data());
+    if (*n_support > 0) {
+      *idx = (int *)allocator->allocate((*n_support) * sizeof(int), stream);
+      MLCommon::copy(*idx, idx_selected.data(), *n_support, stream);
+    }
+  }
+
+  /**
+   * Calculate the b constant in the decision function.
+   *
+   * @param [in] alpha dual coefficients, size [n_rows]
+   * @param [in] f optimality indicator vector, size [n_rows]
+   * @return the value of b
+ */
+  math_t CalcB(const math_t *alpha, const math_t *f) {
+    // We know that for an unbound support vector i, the decision function
+    // (before taking the sign) has value F(x_i) = y_i, where
+    // F(x_i) = \sum_j y_j \alpha_j K(x_j, x_i) + b, and j runs through all
+    // support vectors. The constant b can be expressed from these formulas.
+    // Note that F and f denote different quantities. The lower case f is the
+    // optimality indicator vector defined as
+    // f_i = y_i - \sum_j y_j \alpha_j K(x_j, x_i).
+    // For unbound support vectors f_i = -b.
+
+    // Select f for unbound support vectors (0 < alpha < C)
+    math_t C = this->C;
+    auto select = [C] __device__(math_t a) -> bool { return 0 < a && a < C; };
+    int n_free = SelectByAlpha(alpha, n_rows, f, select, val_selected.data());
+    if (n_free > 0) {
+      cub::DeviceReduce::Sum(cub_storage.data(), cub_bytes, val_selected.data(),
```

Comment:
Not 100% sure but, we should have reduction functions in linalg. Please check the available functions and use the existing one if it works for you.